src/napari/_vispy/overlays/scale_bar.py

```diff
@@ -165,52 +158,32 @@ def _on_data_change(self):
                 color = np.subtract(1, background_color)
                 color[-1] = background_color[-1]
 
-        self.node.set_data(color, self.overlay.ticks)
-        self.node.box.color = box_color
-
-    def _on_box_change(self):
-        self.node.box.visible = self.overlay.box
-
-    def _on_text_change(self):
-        """Update text information"""
-        # update the dpi scale factor to account for screen dpi
-        # because vispy scales pixel height of text by screen dpi
-        if self.node.text.transforms.dpi:
-            # use 96 as the napari reference dpi for historical reasons
-            dpi_scale_factor = 96 / self.node.text.transforms.dpi
-        else:
-            dpi_scale_factor = 1
+        return color, box_color
 
-        self.node.text.font_size = self.overlay.font_size * dpi_scale_factor
-        # changing the fox size changes the box height and positioning in it
-        self.node._update_layout(font_size=self.overlay.font_size)
+    def _on_rendering_change(self):
+        """Change color and other rendering features of scale bar and box."""
+        color, box_color = self._get_colors()
 
-        # positioning in the box uses the center of the box
-        # need to adjust the y_size to be half the size of the current box height
-        self.y_size = self.node.box.height / 2
+        width, height = self.node.set_data(
+            length=self._current_length,
+            color=color,
+            ticks=self.overlay.ticks,
+            font_size=self.overlay.font_size,
+        )
+        self.node.box.color = box_color
 
-        self._on_position_change()
+        self.x_size = width
+        self.y_size = height
 
-    def _on_position_change(self, event=None):
-        # prevent the text from being cut off by shifting down
-        if 'top' in self.overlay.position:
-            # adjust the positioning to account for the box height
-            # 7 is base value for the default 10 font size
-            self.y_offset = 7 + self.y_size
-        else:
-            # ensure if switching from top to bottom, the offset is reset
-            self.y_offset = 7
-        super()._on_position_change()
+    def _on_box_change(self):
+        self.node.box.visible = self.overlay.box
```

Comment:
Better naming? `on_overlay_box_change`? Why assign visible?

Reply:
Why is adding `overlay` better? Thhis is all about the overlay.

This changes whether the box is visible or not, depending on the boolean field of `ScaleBar.box`.

Reply:
It seems to confuse me. When reading this code, I see some generic overlay with a box property, and it is not obvious that it is a boolean flag.

Reply:
Ok. I think that we should rename `overlay` to `overlay_model` (not in this PR)

My confusion comes from overlay having overlay.